modules/web/js/ballerina/components/diagram.jsx

```diff
@@ -58,65 +55,76 @@ class Diagram extends React.Component {
         };
     }
 
+    componentWillMount() {
+        this.visitModel(this.props);
+    }
+
+    componentWillReceiveProps(nextProps) {
+        this.visitModel(nextProps);
+    }
+
+
+    visitModel(nextProps) {
+        try {
+            // 1 , 1.5 and 2 steps has been moved up to ballerina-file-editor.
+            // 3. Now we need to create component for each child of root node.
+            let [others] = [undefined, [], [], []];
+            const otherNodes = [];
+            nextProps.model.children.forEach((child) => {
+                switch (child.constructor.name) {
+                case 'ImportDeclaration':
+                    break;
+                case 'ConstantDefinition':
+                    break;
+                case 'GlobalVariableDefinition':
+                    break;
+                default:
+                    otherNodes.push(child);
+                }
+            });
+            others = getComponentForNodeArray(otherNodes);
+            this.setState({ others });
+            // 3.1 lets filter out annotations so we can overlay html on top of svg.
+            const annotationRenderer = new AnnotationRenderingVisitor();
+            nextProps.model.accept(annotationRenderer);
+            let annotations = [];
+            if (annotationRenderer.getAnnotations()) {
+                annotations = getComponentForNodeArray(annotationRenderer.getAnnotations());
+            }
+            this.setState({ annotations });
+            // 4. Ok we are all set, now lets render the diagram with React. We will create
+            //    s CsnvasDecorator and pass child components for that.
+        } catch (e) {
+            log.error('Error in render calculation.', e);
+            const { editor } = this.context;
+            editor.setActiveView('SOURCE_VIEW');
+        }
+    }
+
     /**
      * @override
      * @memberof Diagram
      */
     render() {
-        // Following is how we render the diagram.
-        // 1. We will visit the model tree and calculate width and height of all
-        //    the elements. We will use DimensionCalcVisitor.
-        this.props.model.accept(this.dimentionCalc);
```

Comment:
@manuranga as we discussed shall we keep the diagram calculations inside the diagram.

As an alternative inside the diagram render we can display an error message instead of rendering the diagram ?